ExternalAHRS: Add the ability to send commands to the EAHRS

valbv opened the pull request:
It's for the ability to send commands to EAHRS like "Enable/Disable GNSS", "Start/Stop magnitometer calibration in flight" and everything that device can receive

Ryanf55 commented:
Nice. If you can, feel free to add one such packet so that this code is used by a driver.

valbv commented:
Linked PR in MAVLink repo:
https://github.com/ArduPilot/mavlink/pull/376

peterbarker reviewed (commented):
I'd prefer to have something at the AHRS-backend level rather than at the EAHRS level for this.  Turning GNSS off is a thing for EKF3, for example, just not in master right now.

valbv commented:
@peterbarker , where can I find the code you were talking about (Turning GNSS off for EKF3)?
I want to avoid conflicts in the names of methods.

I looked through the branches in the repository and requests, but I couldn't find it

Ryanf55 commented:
Check the code path of aux function 65 "GPS Disable": https://ardupilot.org/copter/docs/common-auxiliary-functions.html

valbv commented:
Did I get the idea right, you want to see something like this?

Ryanf55 commented on libraries/AP_AHRS/AP_AHRS.h:
```diff
@@ -683,6 +683,9 @@ class AP_AHRS {
     // get access to an EKFGSF_yaw estimator
     const EKFGSF_yaw *get_yaw_estimator(void) const;
 
+    void set_data_send_disable(bool disable);
```
The name isn't apparent what this does. Can you add a comment? 

Does this disable the EARHS from sending data back to the autopilot?

valbv replied:
I added the comment

valbv replied:
Done

Ryanf55 commented on libraries/AP_ExternalAHRS/AP_ExternalAHRS.h:
```diff
@@ -121,6 +121,9 @@ class AP_ExternalAHRS {
     void send_status_report(class GCS_MAVLINK &link) const;
     bool get_variances(float &velVar, float &posVar, float &hgtVar, Vector3f &magVar, float &tasVar) const;
+    void set_data_send_disable(bool disable);
+    void set_gnss_disable(bool disable);
```
```suggestion
    void set_gnss_disable(bool disable);
```

Does setting this to false re-enable? If so, thoughts on calling it `set_gnss_state(bool enabled)`?

valbv replied:
Yes, I renamed them to

```cpp
void set_data_sending_state(bool enabled);
void set_gnss_state(bool enabled)
```

valbv commented:
Yes, we used `set_gnss_state` method-logic for the dead reckoning tests with EAHRS

Ryanf55 commented on libraries/AP_AHRS/AP_AHRS.cpp:
```diff
@@ -3629,6 +3629,30 @@ float AP_AHRS::get_air_density_ratio(void) const
+#if AP_AHRS_EXTERNAL_ENABLED
+    case EKFType::EXTERNAL:
+        return external.set_gnss_state(enabled);
+#endif
```
The code paths that aux function 105 takes could be used to disable GPS in EKF2/3. 
[105](https://ardupilot.org/copter/docs/common-auxiliary-functions.html)

Not required for the PR.

valbv replied:
Ok, I'll create a next request with these changes

valbv replied:
I'm not sure that I understood what I need right, but I added the GPS disabling for EKF2/3

Ryanf55 commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -4878,6 +4878,39 @@ MAV_RESULT GCS_MAVLINK::handle_command_request_autopilot_capabilities(const mavl
+    switch (packet.command) {
+        case MAV_CMD_AHRS_START:
+        {
+            AP::ahrs().set_data_sending_state(/*enabled*/ true);
```
I know it's ON/OFF, but the fact you need a commen there may be a good reason we can do 

```c++
enum class GNSS_SENDING_STATE {
  ENABLED,
  DISABLED
};
```

valbv replied:
Yeah, I was thinking about that too.
If you don't mind, I'll start a separate next request with the "enum class" using

valbv replied:
Done

Ryanf55 reviewed (approved):
Nice, yea I'm overall on board with the approach here. I haven't tested it yet.

peterbarker commented:
This is looking nicer than the write-bytes approach.

I'm going to mark this for DevCall for discussion.

Ryanf55 commented:
Nothing to add from me, I'm in favor of this change and believe it sets up a nice pattern for adding future generic commands to the EKF/EAHRS.

rmackay9 commented on libraries/AP_ExternalAHRS/AP_ExternalAHRS_InertialLabs.cpp:
```diff
@@ -1120,5 +1128,31 @@ bool AP_ExternalAHRS_InertialLabs::get_variances(float &velVar, float &posVar, f
+
+void AP_ExternalAHRS_InertialLabs::set_data_sending_state(bool enabled)
+{
+    if (enabled)
```
nitpick: bracket should be on the same line as the if

valbv replied:
Done

rmackay9 commented:
Hi @valbv,

Thanks for this.  What's the purpose of disabling an EAHRS?  I guess it's to test a complete failure of the external AHRS system?  I wonder what testing has been done for this?

rmackay9 commented on libraries/AP_AHRS/AP_AHRS_External.cpp:
```diff
@@ -137,4 +137,14 @@ void AP_AHRS_External::get_control_limits(float &ekfGndSpdLimit, float &ekfNavVe
     ekfNavVelGainScaler = 0.5;
 }
 
+void AP_AHRS_External::set_data_sending_state(bool enabled)
```
let's put some comments above the methods..

Ryanf55 replied:
To elaborate, at minimum, copy the comments across all the headers. Although the clasess inherit/override methods, not everyone's editor propogates the description, so AP generally copies comments.

valbv replied:
Done

rmackay9 commented on libraries/AP_AHRS/AP_AHRS_External.h:
```diff
@@ -87,6 +87,9 @@ class AP_AHRS_External : public AP_AHRS_Backend {
 
     void get_control_limits(float &ekfGndSpdLimit, float &controlScaleXY) const override;
+
+    void set_data_sending_state(bool enabled) override;
```
comments here too would be nice

valbv replied:
Done

rmackay9 commented:
More comments would be nice, there are very few which makes it difficult to review but also it will be even more difficult for future developers trying to figure out how to use the various methods

valbv commented on libraries/AP_ExternalAHRS/AP_ExternalAHRS.cpp:
```diff
@@ -375,6 +375,35 @@ void AP_ExternalAHRS::send_status_report(GCS_MAVLINK &link) const
+    }
+    return backend->write_bytes(bytes, len);
+}
+
```
Done

valbv commented on libraries/AP_ExternalAHRS/AP_ExternalAHRS.h:
```diff
@@ -121,6 +121,9 @@ class AP_ExternalAHRS {
     bool get_variances(float &velVar, float &posVar, float &hgtVar, Vector3f &magVar, float &tasVar) const;
+    void set_data_sending_state(bool enabled);
+    void set_gnss_state(bool enabled);
+    bool write_bytes(const char *bytes, uint8_t len);
```
Done

valbv commented on libraries/AP_ExternalAHRS/AP_ExternalAHRS_InertialLabs.h:
```diff
@@ -220,6 +220,9 @@ class AP_ExternalAHRS_InertialLabs : public AP_ExternalAHRS_backend {
     void update_thread();
     bool check_uart();
     bool check_header(const ILabsHeader *h) const;
+    bool write_bytes(const char *bytes, uint8_t len) override;
```
Done

valbv commented on libraries/AP_ExternalAHRS/AP_ExternalAHRS_backend.h:
```diff
@@ -42,6 +42,9 @@ class AP_ExternalAHRS_backend {
     virtual bool get_variances(float &velVar, float &posVar, float &hgtVar, Vector3f &magVar, float &tasVar) const { return false; }
+    virtual void set_data_sending_state(bool enabled) {}
+    virtual void set_gnss_state(bool enabled) {}
+    virtual bool write_bytes(const char *bytes, uint8_t len) { return false; }
```
Done

valbv commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -4878,6 +4878,39 @@ MAV_RESULT GCS_MAVLINK::handle_command_request_autopilot_capabilities(const mavl
+{
+    switch (packet.command) {
+        case MAV_CMD_AHRS_START:
+        {
```
Done

rmackay9 commented:
Are we sure that we want the new methods to be in AP_AHRS?  We could reduce the scope by only making them accessible via AP_ExternalAHRS

Ryanf55 commented:
> Are we sure that we want the new methods to be in AP_AHRS? We could reduce the scope by only making them accessible via AP_ExternalAHRS

Reasoning: https://github.com/ArduPilot/ardupilot/pull/28485#pullrequestreview-2486838660

rmackay9 commented:
I wonder if we even really need the mavlink command?  Some of our other subsystems can be enabled/disabled using a _TYPE parameter.  If it's not something that users will regularly do a parameter can be more convenient and accessible (most GCSs won't support this mavlink command I suspect)

valbv commented:
If this PR will be Ok, I plan to add common commands for AHRS In the next PRs:
- Start/Stop magnetometer calibration
- Send aiding data

That's a reason why I chose the MAVLink protocol extension.
And then add the "AHRS" tab to the Mission Planner

tridge commented:
@valbv for this type of runtime configuration we have in the past used aux functions, which are available over mavlink as well as R/C switches and lua scripts. Is there some reason for not continuing with that approach? 
I also wouldn't mind having a COMMAND_INT that is reserved for external AHRS, where the interpretation of the params is up to each backend. That would give a lot of flexibility if aux functions aren't sufficient

valbv commented:
@tridge, sorry for the long absence.
Aux functions seem to be quite sufficient for this functionality.
I removed changes in MAVLink and added RC-Channel processing for it.

peterbarker commented on libraries/AP_AHRS/AP_AHRS.cpp:
```diff
@@ -3652,6 +3652,50 @@ float AP_AHRS::get_air_density_ratio(void) const
     return 1.0 / sq(eas2tas);
 }
 
+void AP_AHRS::set_data_sending_state(DATA_SENDING_STATE data_sending_state)
```
```suggestion
void AP_AHRS::set_data_sending_state(DataSendingState new_state)
```

similarly for other enum names

valbv replied:
Done

peterbarker commented on libraries/AP_AHRS/AP_AHRS_Backend.h:
```diff
@@ -292,4 +292,17 @@ class AP_AHRS_Backend
+
+    enum class DATA_SENDING_STATE {
+        ENABLED,
+        DISABLED
```
```suggestion
        DISABLED,
```

similarly elsewhere

valbv replied:
Done

peterbarker commented on libraries/AP_AHRS/AP_AHRS_Backend.h:
```diff
@@ -292,4 +292,17 @@ class AP_AHRS_Backend
+
+    enum class GPS_STATE {
+        ENABLED,
+        DISABLED
```
```suggestion
        DISABLED,
```

valbv replied:
Done

peterbarker commented on libraries/AP_AHRS/AP_AHRS.cpp:
```diff
@@ -3652,6 +3652,50 @@ float AP_AHRS::get_air_density_ratio(void) const
+
+void AP_AHRS::set_gps_state(GPS_STATE gps_state)
+{
+    switch (active_EKF_type()) {
```
This is somewhat odd.  You're turning things off the for active EKLF type.  Presumably it's not going to stay the active EKF type for long, so turning the functionality back on again could be rather problematic?

valbv replied:
Fixed

peterbarker commented on libraries/AP_AHRS/AP_AHRS.cpp:
```diff
@@ -3652,6 +3652,50 @@ float AP_AHRS::get_air_density_ratio(void) const
+                                      AP_AHRS_Backend::GPS_STATE::DISABLED :
+                                      AP_AHRS_Backend::GPS_STATE::ENABLED);
+#endif
+    default:
+        break;
```
Avoid the use of default: clauses, spell out each backend type

valbv replied:
Done

peterbarker commented on libraries/AP_AHRS/AP_AHRS_External.cpp:
```diff
@@ -137,4 +137,18 @@ void AP_AHRS_External::get_control_limits(float &ekfGndSpdLimit, float &ekfNavVe
 
+void AP_AHRS_External::set_data_sending_state(DATA_SENDING_STATE state)
+{
+    AP::externalAHRS().set_data_sending_state(state == DATA_SENDING_STATE::ENABLED ?
+                                              AP_ExternalAHRS::DATA_SENDING_STATE::ENABLED :
+                                              AP_ExternalAHRS::DATA_SENDING_STATE::DISABLED);
```
```suggestion
    const bool ext_state = (state == DataSendingState::ENABLED) ?
                                              AP_ExternalAHRS::DataSendingState::ENABLED :
                                              AP_ExternalAHRS::DataSendingState::DISABLED
    AP::externalAHRS().set_data_sending_state(ext_state);
```

similarly elsewhere

valbv replied:
Done

peterbarker reviewed (commented):
I'm marking this for DevCall.

Looking at this, I think the AHRS interface is problematic and will remain so until we clean up the AHRS backend stuff properly.

I'm thinking one of two things:
 - explicit aux functions for disabling stuff in ExternalAHRS (ie. calls directly on the EAHRS library
 - passing in a "dummy" instance number for the AHRS enable/disable calls, 0 for primary, 1 for secondary estimators.  You'd need separate functions for each.
 -

IamPete1 commented on libraries/AP_AHRS/AP_AHRS.cpp:
```diff
@@ -3652,6 +3652,34 @@ float AP_AHRS::get_air_density_ratio(void) const
+    external.set_gps_state(ext_state);
+#endif
+
+#if HAL_NAVEKF2_AVAILABLE && AP_GPS_ENABLED
+    AP::gps().force_disable(new_state == GpsState::DISABLED);
+#endif
+
+#if HAL_NAVEKF3_AVAILABLE && AP_GPS_ENABLED
+    AP::gps().force_disable(new_state == GpsState::DISABLED);
+#endif
```
Don't need this twice.

valbv replied:
Sorry, I don't understand what you mean. Can you rephrase it, please?

Here are cases for EKF2 and EKF3

IamPete1 replied:
Right, but you can combine the defines. 
```
#if (HAL_NAVEKF2_AVAILABLE || HAL_NAVEKF3_AVAILABLE) && AP_GPS_ENABLED
```

No need to duplicate the function call.

valbv replied:
Done

IamPete1 commented on libraries/AP_AHRS/AP_AHRS.cpp:
```diff
@@ -3652,6 +3652,34 @@ float AP_AHRS::get_air_density_ratio(void) const
+#endif
+}
+
+void AP_AHRS::set_gps_state(GpsState new_state)
```
Is this method used?

valbv replied:
Yes, in the "RC_Channel.cpp"
